src/core/lib/config/core_configuration.h

```diff
@@ -0,0 +1,74 @@
+// Copyright 2021 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GRPC_CORE_LIB_CONFIG_CORE_CONFIGURATION_H
+#define GRPC_CORE_LIB_CONFIG_CORE_CONFIGURATION_H
+
+#include <grpc/impl/codegen/port_platform.h>
+
+#include <atomic>
+
+namespace grpc_core {
+
+// Global singleton that stores library configuration - factories, etc...
+// that plugins might choose to extend.
+class CoreConfiguration {
+ public:
+  CoreConfiguration(const CoreConfiguration&) = delete;
+  CoreConfiguration& operator=(const CoreConfiguration&) = delete;
+
+  // Builder is passed to plugins, etc... at initialization time to collect
+  // their configuration and assemble the published CoreConfiguration.
+  class Builder {
```

Comment:
It seems like we have two levels of indirection here, one for the `Builder` class and another for the `BuildCoreConfiguration()` function.  What's the purpose of all of this indirection?  Aren't we ultimately going to need all state associated with plugins to be stored in `CoreConfiguration` directly?

To say this another way, it seems like in order to add any plugin, we're going to plumb its state through all three places.  Wouldn't it make more sense to just have a single object (`CoreConfiguration`) in which to store that state?

Reply:
Plugin's won't generally add kinds of state to CoreConfiguration, systems will (like a channel, or a resolver).

There's a circular dependence in the whole initialization scheme that we have now, that's preserved here: the initialization system needs to make an upcall to the thing that knows about all the plugins, so that they can make downcalls into the initialization system. That's the purpose of `BuildCoreConfiguration` - it's the global plugin init function of the 2020's. (note: if we switched to dynamic registration of plugins via constructors, we'd still want a function with the role of BuildCoreConfiguration to read through the list of things that should be registered - I think this function's role is fundamental).

The question then comes: what interface should be presented to BuildCoreConfiguration, and there are two choices:
1. `CoreConfiguration*` - we could have mutable methods on CoreConfiguration to add things to it, and const functions to read from it, and have the convention that we only mutate at initialization time
2. `CoreConfiguration::Builder*` - we create a separate object to locate initialization methods on, more strictly defining the phases of usage: there is a build time, and then there is a usage time, and they are expressed via two different interfaces. This also gives us a place to locate configuration finalization, of which we already have some examples: https://github.com/grpc/grpc/blob/master/src/core/lib/surface/channel_init.cc#L77.

Since we have use cases for 2, and it more clearly expresses the semantics we want, and it's rare that we add a system that needs a new kind of configuration (so a little extra typing is ok), I went with the `Builder` based API.

Reply:
Thanks for explaining.  This all makes sense.